[JSC] Inline ClonedArguments allocation in DFG / FTL

Constellation opened the pull request:
#### c7d1888949f94118612536ffc3b7f58cf102114b
<pre>
[JSC] Inline ClonedArguments allocation in DFG / FTL
<a href="https://bugs.webkit.org/show_bug.cgi?id=256129">https://bugs.webkit.org/show_bug.cgi?id=256129</a>
rdar://108694519

Reviewed by Mark Lam and Keith Miller.

This patch adds ClonedArguments inlined allocation into DFG / FTL.
ClonedArguments is `arguments` object in strict mode, and we should make it super fast.
The approach is just allocating butterfly and object from JIT.

                                        ToT                     Patched

    create-cloned-arguments       17.9729+-0.3882     ^      7.8421+-0.0352        ^ definitely 2.2919x faster

* JSTests/microbenchmarks/create-cloned-arguments.js: Added.
(test):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compileCreateClonedArguments):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compileCreateClonedArguments):
* Source/JavaScriptCore/ftl/FTLAbstractHeapRepository.cpp:
* Source/JavaScriptCore/ftl/FTLAbstractHeapRepository.h:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCreateClonedArguments):
* Source/JavaScriptCore/jit/AssemblyHelpers.h:
(JSC::AssemblyHelpers::emitFillStorageWithJSEmpty):
* Source/JavaScriptCore/runtime/ClonedArguments.cpp:
(JSC::ClonedArguments::getOwnPropertySlot):
* Source/JavaScriptCore/runtime/ClonedArguments.h:

Canonical link: <a href="https://commits.webkit.org/263558@main">https://commits.webkit.org/263558@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5ab953dfb91e3c27a17985804b2b375d9d5070a0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5058 "12 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5189 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5365 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6578 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5148 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5384 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5171 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/5397 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5144 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5230 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4537 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6594 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2717 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4535 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/9576 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/4187 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4580 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4608 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6218 "Passed tests") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/4660 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5023 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/4126 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/5151 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4509 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1296 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8586 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/5295 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/568 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4875 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1411 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

kmiller68 reviewed (approved):
r=me.

MenloDorian commented on Source/JavaScriptCore/runtime/ClonedArguments.cpp:
```diff
@@ -93,6 +85,15 @@ ClonedArguments* ClonedArguments::createWithInlineFrame(JSGlobalObject* globalOb
 
     ClonedArguments* result = nullptr;
+
+    auto createEmptyWithCheck = [&](JSGlobalObject* globalObject, JSFunction* callee, unsigned length) {
```
I presume by "Check", you mean the ASSERT?  How about naming this `createEmptyWithAssert` instead?

Constellation replied:
Changed.

MenloDorian commented on Source/JavaScriptCore/runtime/ClonedArguments.cpp:
```diff
@@ -40,24 +40,26 @@ ClonedArguments::ClonedArguments(VM& vm, Structure* structure, Butterfly* butter
+            indexingHeader.setVectorLength(vectorLength);
+            indexingHeader.setPublicLength(length);
+            butterfly = Butterfly::tryCreate(vm, nullptr, 0, structure->outOfLineCapacity(), true, indexingHeader, vectorLength * sizeof(EncodedJSValue));
+            if (!butterfly)
+                return nullptr;
```
Should we ever check and return `nullptr` here instead of a RELEASE_ASSERT?  It seems the client below assumes that it never fails to allocate.  I guess it's ok to leave as is for now, but I suspect we should change this to a RELEASE_ASSERT later.

Constellation replied:
For now, I'm not changing it since it is the same to the current behavior.

MenloDorian commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
```diff
@@ -6823,6 +6824,121 @@ void SpeculativeJIT::compileEnumeratorPutByVal(Node* node)
+            static_assert((1U << 3) == sizeof(JSValue));
+            lshift32(sizeGPR, TrustedImm32(3), scratchGPR);
+            add32(TrustedImm32(sizeof(IndexingHeader) + outOfLineCapacity * sizeof(JSValue)), scratchGPR, scratch2GPR);
+            move(TrustedImmPtr(nullptr), storageGPR);
```
Do we really need to set `storageGPR` to nullptr here given that `emitAllocateVariableSized` is going to set it immediately below?

Oh, I see ... the `slowCases` path relies on `storageGPR` being initialized.  However, you have a branch above for the `sizeGPR` vs `MAX_STORAGE_VECTOR_LENGTH` comparison.  I suppose it doesn't matter because the slow path operation will do a length check first.  However, it does feel a little brittle to pass an uninitialized pointer like this.  Can we move this initialization of `storageGPR` above the branch above?

Constellation replied:
Moved.

MenloDorian commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -7607,13 +7608,68 @@ IGNORE_CLANG_WARNINGS_END
+            LValue length = getArgumentsLength().value;
+            LValue argumentRegion = getArgumentsStart();
+            LValue callee = getCurrentCallee();
+            LValue byteSize = m_out.add(m_out.constIntPtr(sizeof(IndexingHeader) + outOfLineCapacity * sizeof(JSValue)), m_out.shl(m_out.zeroExtPtr(length), m_out.constInt32(3)));
```
Why do we not need the length check against `MAX_STORAGE_VECTOR_LENGTH` up top like we do in the DFG?

Constellation replied:
Added.

MenloDorian commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -7607,13 +7608,68 @@ IGNORE_CLANG_WARNINGS_END
+            for (unsigned i = 0; i < outOfLineCapacity; ++i)
+                m_out.store64(m_out.int64Zero, m_out.address(m_heaps.properties.atAnyNumber(), storage, i * sizeof(JSValue)));
+            LValue butterfly = m_out.add(m_out.constIntPtr(sizeof(IndexingHeader) + outOfLineCapacity * sizeof(JSValue)), storage);
+            m_out.store32(length, butterfly, m_heaps.Butterfly_publicLength);
+            m_out.store32(length, butterfly, m_heaps.Butterfly_vectorLength);
```
Is there no storePair idiom in the B3?  Or will the B3 backend collate this into a storePair?

Constellation replied:
B3 does not have this.

MenloDorian reviewed (approved):
r=me

Constellation committed:
[JSC] Inline ClonedArguments allocation in DFG / FTL
https://bugs.webkit.org/show_bug.cgi?id=256129
rdar://108694519

Reviewed by Mark Lam and Keith Miller.

This patch adds ClonedArguments inlined allocation into DFG / FTL.
ClonedArguments is `arguments` object in strict mode, and we should make it super fast.
The approach is just allocating butterfly and object from JIT.

                                        ToT                     Patched

    create-cloned-arguments       17.9729+-0.3882     ^      7.8421+-0.0352        ^ definitely 2.2919x faster

* JSTests/microbenchmarks/create-cloned-arguments.js: Added.
(test):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compileCreateClonedArguments):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compileCreateClonedArguments):
* Source/JavaScriptCore/ftl/FTLAbstractHeapRepository.cpp:
* Source/JavaScriptCore/ftl/FTLAbstractHeapRepository.h:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCreateClonedArguments):
* Source/JavaScriptCore/jit/AssemblyHelpers.h:
(JSC::AssemblyHelpers::emitFillStorageWithJSEmpty):
* Source/JavaScriptCore/runtime/ClonedArguments.cpp:
(JSC::ClonedArguments::getOwnPropertySlot):
* Source/JavaScriptCore/runtime/ClonedArguments.h:

Canonical link: https://commits.webkit.org/263558@main
